File: packages/python/src/brainpick/compile/t2.py
"""T2: the vectors tier (spec/30) — deterministic chunking, embedding, LanceDB.

The chunker is normative: both engines must produce byte-identical
t2/chunks.jsonl. Everything here is char-based — no tokenizer dependency.
"""
from __future__ import annotations

import json
import os
import re
from dataclasses import dataclass
from pathlib import Path

from brainpick.core.canonical import canonical_json, canonical_jsonl, sha256_hex
from brainpick.core.fs import write_if_changed
from brainpick.embed import make_embedder
from brainpick.vectorstore import VectorStore, lancedb_available

MAX_CHUNK = 3200   # chars, hard budget per chunk (overlap counts toward it)
OVERLAP = 320      # chars of the previous chunk prefixed onto the next

_HEADING = re.compile(r"^(#{1,3}) (.*)$")
_FENCE = re.compile(r"^(`{3,}|~{3,})")
_SLUG_RUNS = re.compile(r"[\W_]+", re.UNICODE)  # non-alphanumeric runs (spec/30: `_` too)


def slugify(title: str) -> str:
    """lowercase; every run of non-alphanumerics → `-`; trimmed of `-` (spec/30)."""
    return _SLUG_RUNS.sub("-", title.lower()).strip("-")


def _fence_close(opening: str):
    """A fence closes on the same char, at least as long, nothing else on the line."""
    return re.compile(rf"^{re.escape(opening[0])}{{{len(opening)},}}\s*$")


def _split_sections(text: str) -> list[tuple[list[str], list[str]]]:
    """[(heading_path, content lines)] — split at ATX headings 1–3 outside fences.

    The heading line itself is not part of the section's content: the titles
    travel in heading_path (metadata), the text stays prose.
    """
    levels: dict[int, str] = {}
    sections: list[tuple[list[str], list[str]]] = [([], [])]
    fence_close = None
    for line in text.split("\n"):
        if fence_close is not None:
            sections[-1][1].append(line)
            if fence_close.match(line):
                fence_close = None
            continue
        fence = _FENCE.match(line)
        if fence:
            fence_close = _fence_close(fence.group(1))
            sections[-1][1].append(line)
            continue
        heading = _HEADING.match(line)
        if heading:
            level = len(heading.group(1))
            levels = {lv: t for lv, t in levels.items() if lv < level}
            levels[level] = heading.group(2).strip()
            path = [levels[lv] for lv in sorted(levels)]
            sections.append((path, []))
            continue
        sections[-1][1].append(line)
    return sections


def _split_paragraphs(lines: list[str]) -> list[str]:
    """Blank-line separated paragraphs; blank lines inside fences do not split."""
    paragraphs: list[str] = []
    current: list[str] = []
    fence_close = None
    for line in lines:
        if fence_close is not None:
            current.append(line)
            if fence_close.match(line):
                fence_close = None
            continue
        fence = _FENCE.match(line)
        if fence:
            fence_close = _fence_close(fence.group(1))
            current.append(line)
            continue
        if line.strip() == "":
            if current:
                paragraphs.append("\n".join(current))
                current = []
            continue
        current.append(line)
    if current:
        paragraphs.append("\n".join(current))
    return paragraphs


def _pack(paragraphs: list[str]) -> list[str]:
    """Greedy packing to MAX_CHUNK; chunks after the first reserve OVERLAP chars
    for the incoming prefix; a paragraph over the budget is hard-split."""
    chunks: list[str] = []
    parts: list[str] = []

    def budget() -> int:
        return MAX_CHUNK if not chunks else MAX_CHUNK - OVERLAP

    for paragraph in paragraphs:
        if parts and len("\n\n".join([*parts, paragraph])) > budget():
            chunks.append("\n\n".join(parts))
            parts = []
        if not parts and len(paragraph) > budget():
            rest = paragraph
            while len(rest) > budget():
                cut = budget()  # 3200 for a section's first chunk, 2880 after
                chunks.append(rest[:cut])
                rest = rest[cut:]
            parts = [rest] if rest else []
        else:
            parts.append(paragraph)
    if parts:
        chunks.append("\n\n".join(parts))
    return chunks


def chunk_document(record: dict) -> list[dict]:
    """spec/30's normative chunker over one docs.jsonl record (never a reserved doc)."""
    doc = record["path"]
    result: list[dict] = []
    ord_counter = 0
    for heading_path, lines in _split_sections(record["text"]):
        base = _pack(_split_paragraphs(lines))
        emitted: list[str] = []
        for i, text in enumerate(base):
            emitted.append(text if i == 0 else emitted[-1][-OVERLAP:] + text)
        slug_path = "/".join(slugify(title) for title in heading_path)
        n = 0
        for text in emitted:
            if not text.strip():
                continue  # spec/30 step 4: empty or whitespace-only chunks are dropped
            result.append({
                "doc": doc,
                "heading_path": list(heading_path),
                "id": f"{doc}#{slug_path}~{n}",
                "ord": ord_counter,
                "sha256": sha256_hex(text.encode("utf-8")),
                "text": text,
            })
            n += 1
            ord_counter += 1
    return result


def build_chunks(records: list[dict]) -> list[dict]:
    """Every chunk of every non-reserved document, sorted by (doc, ord)."""
    chunks: list[dict] = []
    for record in sorted(records, key=lambda r: r["path"]):
        if record["reserved"]:
            continue
        chunks.extend(chunk_document(record))
    return chunks


# -- the compile stage ---------------------------------------------------------------


@dataclass
class T2Result:
    status: str            # "fresh" | "stale" — tiers.t2 for the manifest
    changed: bool          # chunks.jsonl or embedding.json bytes changed
    warning: str | None = None


def fingerprint(kind: str, endpoint: str, model: str, dim: int,
                document_prefix: str = "", query_prefix: str = "") -> str:
    """First 16 hex chars of sha256 over `kind|endpoint|model|dim` (spec/30) —
    `|document_prefix|query_prefix` appended only when a prefix is set."""
    key = f"{kind}|{endpoint}|{model}|{dim}"
    if document_prefix or query_prefix:
        key += f"|{document_prefix}|{query_prefix}"
    return sha256_hex(key.encode("utf-8"))[:16]


def t2_gate(config) -> tuple[bool, str | None]:
    """(enabled, instruction) per [modules] vectors: auto lights up when an
    embedding backend is configured AND lancedb is importable; otherwise the
    instruction names exactly what is missing (spec/30 detection ladder, rung 6)."""
    if config.modules.vectors == "off":
        return False, None  # the user chose off — no nagging
    if not config.models.embedding.kind:
        return False, (
            "T2 vectors off — no [models.embedding] in brainpick.toml; "
            "`brainpick init` detects local backends (ollama pull nomic-embed-text)"
        )
    if not lancedb_available():
        return False, (
            "T2 vectors off — the vector store is missing: pip install 'brainpick[vectors]'"
        )
    return True, None


def _normalized_backend(embedding) -> tuple[str, str, str]:
    """Config → the (kind, endpoint, model) recorded in embedding.json.

    `openai` (what init records for the paid API) is an openai-compatible
    endpoint; embedding.json keeps the spec/30 enum."""
    kind = "openai-compatible" if embedding.kind == "openai" else embedding.kind
    model = embedding.model or ("mock" if kind == "mock" else "")
    return kind, embedding.endpoint, model


def run_t2_stage(bp: Path, records: list[dict], embedding, full: bool = False) -> T2Result:
    """Compile chunks + vectors under <bp>/t2; never raises (failures degrade)."""
    chunks = build_chunks(records)
    chunks_changed = write_if_changed(bp / "t2" / "chunks.jsonl", canonical_jsonl(chunks))
    try:
        embedding_changed = _sync_vectors(bp, chunks, embedding, full)
    except Exception as error:  # T2 failures never block T1 (spec/00 degradation ladder)
        return T2Result("stale", chunks_changed, warning=(
            f"T2 embedding failed ({error}) — semantic search degrades to keyword; "
            "fix the backend and recompile"
        ))
    return T2Result("fresh", chunks_changed or embedding_changed)


def _read_json(path: Path) -> dict | None:
    if not path.is_file():
        return None
    try:
        return json.loads(path.read_text(encoding="utf-8"))
    except (OSError, ValueError):
        return None


def _sync_vectors(bp: Path, chunks: list[dict], embedding, full: bool) -> bool:
    kind, endpoint, model = _normalized_backend(embedding)
    embedder = make_embedder(kind, endpoint, model, api_key=os.environ.get("OPENAI_API_KEY", ""))
    store = VectorStore(bp / "t2" / "lancedb")

    doc_prefix, query_prefix = embedding.document_prefix, embedding.query_prefix
    old = _read_json(bp / "t2" / "embedding.json")
    same_backend = (
        not full
        and old is not None
        and (old.get("kind"), old.get("endpoint"), old.get("model")) == (kind, endpoint, model)
        and (old.get("document_prefix", ""), old.get("query_prefix", "")) == (doc_prefix, query_prefix)
    )

    if same_backend:
        embedded = store.existing_shas()
        new_ids = {chunk["id"] for chunk in chunks}
        to_embed = [c for c in chunks if embedded.get(c["id"]) != c["sha256"]]
        delete_ids = (set(embedded) - new_ids) | {c["id"] for c in to_embed if c["id"] in embedded}
    else:
        to_embed, delete_ids = chunks, set()

    vectors = embedder.embed([doc_prefix + chunk["text"] for chunk in to_embed]) if to_embed else []
    if to_embed:
        dim = len(vectors[0])
    elif same_backend:
        dim = int(old["dim"])
    elif embedding.dim:
        dim = int(embedding.dim)
    else:
        dim = len(embedder.embed(["brainpick"])[0])  # discover once, even with no chunks

    if same_backend and dim != int(old["dim"]):
        # the backend answers with a new dimensionality — every old vector is invalid
        same_backend = False
        to_embed = chunks
        vectors = embedder.embed([doc_prefix + chunk["text"] for chunk in chunks])

    rows = [
        {"id": c["id"], "doc": c["doc"], "ord": c["ord"], "text": c["text"], "vector": v}
        for c, v in zip(to_embed, vectors)
    ]
    if same_backend:
        if rows or delete_ids:
            store.upsert(rows, delete_ids, dim)
    else:
        store.replace_all(rows, dim)

    record = {
        "dim": dim,
        "endpoint": endpoint,
        "fingerprint": fingerprint(kind, endpoint, model, dim, doc_prefix, query_prefix),
        "kind": kind,
        "model": model,
    }
    if doc_prefix or query_prefix:  # spec/30: keys appear only when a prefix is set
        record["document_prefix"] = doc_prefix
        record["query_prefix"] = query_prefix
    return write_if_changed(bp / "t2" / "embedding.json", canonical_json(record))

File: packages/node/src/compile/t2.ts
/** T2: the vectors tier (spec/30) — deterministic chunking, embedding, LanceDB.
 *
 * The chunker is normative: both engines must produce byte-identical
 * t2/chunks.jsonl. Everything here is char-based — no tokenizer dependency —
 * where "char" means Unicode code point (Python `len`/slicing), NOT UTF-16
 * code units; the cp* helpers below carry that difference.
 */
import { join } from "node:path";

import type { Config, EmbeddingConfig } from "../config";
import { canonicalJson, canonicalJsonl, cmpStr, sha256Hex, type JsonValue } from "../core/canonical";
import { readTextOrNull, writeIfChanged } from "../core/fs";
import { PY_SPACE_CLASS, pyStrip } from "../core/pyfmt";
import { DEFAULT_LOCAL_MODEL, makeEmbedder } from "../embed";
import { lancedbAvailable, VectorStore, type ChunkRow } from "../vectorstore";

export const MAX_CHUNK = 3200; // chars, hard budget per chunk (overlap counts toward it)
export const OVERLAP = 320; // chars of the previous chunk prefixed onto the next

// `[^\n]*$` instead of `.*$`: unlike Python's `.`, JS `.` refuses `\r`, which
// would unmatch headings in CRLF bodies.
const HEADING = /^(#{1,3}) ([^\n]*)$/;
const FENCE = /^(`{3,}|~{3,})/;
const SLUG_RUNS = /[^\p{L}\p{N}]+/gu; // non-alphanumeric runs (spec/30: `_` too)

/** lowercase; every run of non-alphanumerics → `-`; trimmed of `-` (spec/30). */
export function slugify(title: string): string {
  return title.toLowerCase().replace(SLUG_RUNS, "-").replace(/^-+|-+$/g, "");
}

/** A fence closes on the same char, at least as long, nothing else on the line. */
function fenceCloseRe(opening: string): RegExp {
  return new RegExp(`^${opening[0]}{${opening.length},}[${PY_SPACE_CLASS}]*$`, "u");
}

/** Python `len(s)` — code points, not UTF-16 units. */
function cpLen(s: string): number {
  let n = 0;
  for (const _ of s) n++;
  return n;
}

/** Python `s[-n:]` — the last n code points. */
function cpTail(s: string, n: number): string {
  const arr = [...s];
  return arr.length <= n ? s : arr.slice(arr.length - n).join("");
}

/** Python `s[:n]` — the first n code points. */
export function cpHead(s: string, n: number): string {
  const arr = [...s];
  return arr.length <= n ? s : arr.slice(0, n).join("");
}

interface Section {
  headingPath: string[];
  lines: string[];
}

/** [(heading_path, content lines)] — split at ATX headings 1–3 outside fences.
 *
 * The heading line itself is not part of the section's content: the titles
 * travel in heading_path (metadata), the text stays prose. */
function splitSections(text: string): Section[] {
  const levels = new Map<number, string>();
  const sections: Section[] = [{ headingPath: [], lines: [] }];
  let fenceClose: RegExp | null = null;
  for (const line of text.split("\n")) {
    if (fenceClose !== null) {
      sections[sections.length - 1]!.lines.push(line);
      if (fenceClose.test(line)) fenceClose = null;
      continue;
    }
    const fence = FENCE.exec(line);
    if (fence) {
      fenceClose = fenceCloseRe(fence[1]!);
      sections[sections.length - 1]!.lines.push(line);
      continue;
    }
    const heading = HEADING.exec(line);
    if (heading) {
      const level = heading[1]!.length;
      for (const lv of [...levels.keys()]) {
        if (lv >= level) levels.delete(lv);
      }
      levels.set(level, pyStrip(heading[2]!));
      const path = [...levels.keys()].sort((a, b) => a - b).map((lv) => levels.get(lv)!);
      sections.push({ headingPath: path, lines: [] });
      continue;
    }
    sections[sections.length - 1]!.lines.push(line);
  }
  return sections;
}

/** Blank-line separated paragraphs; blank lines inside fences do not split. */
function splitParagraphs(lines: string[]): string[] {
  const paragraphs: string[] = [];
  let current: string[] = [];
  let fenceClose: RegExp | null = null;
  for (const line of lines) {
    if (fenceClose !== null) {
      current.push(line);
      if (fenceClose.test(line)) fenceClose = null;
      continue;
    }
    const fence = FENCE.exec(line);
    if (fence) {
      fenceClose = fenceCloseRe(fence[1]!);
      current.push(line);
      continue;
    }
    if (pyStrip(line) === "") {
      if (current.length > 0) {
        paragraphs.push(current.join("\n"));
        current = [];
      }
      continue;
    }
    current.push(line);
  }
  if (current.length > 0) paragraphs.push(current.join("\n"));
  return paragraphs;
}

/** Greedy packing to MAX_CHUNK; chunks after the first reserve OVERLAP chars
 * for the incoming prefix; a paragraph over the budget is hard-split. */
function pack(paragraphs: string[]): string[] {
  const chunks: string[] = [];
  let parts: string[] = [];

  const budget = (): number => (chunks.length === 0 ? MAX_CHUNK : MAX_CHUNK - OVERLAP);

  for (const paragraph of paragraphs) {
    if (parts.length > 0 && cpLen([...parts, paragraph].join("\n\n")) > budget()) {
      chunks.push(parts.join("\n\n"));
      parts = [];
    }
    if (parts.length === 0 && cpLen(paragraph) > budget()) {
      let rest = [...paragraph]; // code points
      while (rest.length > budget()) {
        const cut = budget(); // 3200 for a section's first chunk, 2880 after
        chunks.push(rest.slice(0, cut).join(""));
        rest = rest.slice(cut);
      }
      parts = rest.length > 0 ? [rest.join("")] : [];
    } else {
      parts.push(paragraph);
    }
  }
  if (parts.length > 0) chunks.push(parts.join("\n\n"));
  return chunks;
}

export interface Chunk {
  doc: string;
  heading_path: string[];
  id: string;
  ord: number;
  sha256: string;
  text: string;
}

export interface ChunkSource {
  path: string;
  text: string;
  reserved: boolean;
}

/** spec/30's normative chunker over one docs.jsonl record (never a reserved doc). */
export function chunkDocument(record: Pick<ChunkSource, "path" | "text">): Chunk[] {
  const doc = record.path;
  const result: Chunk[] = [];
  let ordCounter = 0;
  for (const { headingPath, lines } of splitSections(record.text)) {
    const base = pack(splitParagraphs(lines));
    const emitted: string[] = [];
    for (let i = 0; i < base.length; i++) {
      emitted.push(i === 0 ? base[i]! : cpTail(emitted[emitted.length - 1]!, OVERLAP) + base[i]!);
    }
    const slugPath = headingPath.map(slugify).join("/");
    let n = 0;
    for (const text of emitted) {
      if (pyStrip(text) === "") continue; // spec/30 step 4: blank chunks are dropped
      result.push({
        doc,
        heading_path: [...headingPath],
        id: `${doc}#${slugPath}~${n}`,
        ord: ordCounter,
        sha256: sha256Hex(text),
        text,
      });
      n += 1;
      ordCounter += 1;
    }
  }
  return result;
}

/** Every chunk of every non-reserved document, sorted by (doc, ord). */
export function buildChunks(records: readonly ChunkSource[]): Chunk[] {
  const chunks: Chunk[] = [];
  for (const record of [...records].sort((a, b) => cmpStr(a.path, b.path))) {
    if (record.reserved) continue;
    chunks.push(...chunkDocument(record));
  }
  return chunks;
}

// -- the compile stage ---------------------------------------------------------------

export interface T2Result {
  status: "fresh" | "stale"; // tiers.t2 for the manifest
  changed: boolean; // chunks.jsonl or embedding.json bytes changed
  warning: string | null;
}

/** First 16 hex chars of sha256 over `kind|endpoint|model|dim` (spec/30). */
export function fingerprint(
  kind: string,
  endpoint: string,
  model: string,
  dim: number,
  documentPrefix = "",
  queryPrefix = "",
): string {
  // spec/30: `|document_prefix|query_prefix` appended only when a prefix is set
  let key = `${kind}|${endpoint}|${model}|${dim}`;
  if (documentPrefix || queryPrefix) key += `|${documentPrefix}|${queryPrefix}`;
  return sha256Hex(key).slice(0, 16);
}

/** (enabled, instruction) per [modules] vectors: auto lights up when an
 * embedding backend is configured AND lancedb is importable; otherwise the
 * instruction names exactly what is missing (spec/30 detection ladder, rung 6). */
export async function t2Gate(config: Config): Promise<[boolean, string | null]> {
  if (config.modules.vectors === "off") return [false, null]; // the user chose off — no nagging
  if (!config.models.embedding.kind) {
    return [
      false,
      "T2 vectors off — no [models.embedding] in brainpick.toml; " +
        "`brainpick init` detects local backends (ollama pull nomic-embed-text)",
    ];
  }
  if (!(await lancedbAvailable())) {
    return [
      false,
      "T2 vectors off — the vector store is missing: npm install @lancedb/lancedb",
    ];
  }
  return [true, null];
}

/** Config → the (kind, endpoint, model) recorded in embedding.json.
 *
 * `openai` (what init records for the paid API) is an openai-compatible
 * endpoint; embedding.json keeps the spec/30 enum. */
function normalizedBackend(embedding: EmbeddingConfig): [string, string, string] {
  const kind = embedding.kind === "openai" ? "openai-compatible" : embedding.kind;
  const model = embedding.model || (kind === "mock" ? "mock" : kind === "local" ? DEFAULT_LOCAL_MODEL : "");
  return [kind, embedding.endpoint, model];
}

/** Compile chunks + vectors under <bp>/t2; never throws (failures degrade). */
export async function runT2Stage(
  bp: string,
  records: readonly ChunkSource[],
  embedding: EmbeddingConfig,
  full = false,
): Promise<T2Result> {
  const chunks = buildChunks(records);
  const chunksChanged = writeIfChanged(
    join(bp, "t2", "chunks.jsonl"),
    canonicalJsonl(chunks as unknown as JsonValue[]),
  );
  let embeddingChanged: boolean;
  try {
    embeddingChanged = await syncVectors(bp, chunks, embedding, full);
  } catch (error) {
    // T2 failures never block T1 (spec/00 degradation ladder)
    const msg = error instanceof Error ? error.message : String(error);
    return {
      status: "stale",
      changed: chunksChanged,
      warning:
        `T2 embedding failed (${msg}) — semantic search degrades to keyword; ` +
        "fix the backend and recompile",
    };
  }
  return { status: "fresh", changed: chunksChanged || embeddingChanged, warning: null };
}

function readJson(path: string): Record<string, unknown> | null {
  const text = readTextOrNull(path);
  if (text === null) return null;
  try {
    const parsed: unknown = JSON.parse(text);
    if (typeof parsed !== "object" || parsed === null || Array.isArray(parsed)) return null;
    return parsed as Record<string, unknown>;
  } catch {
    return null;
  }
}

async function syncVectors(
  bp: string,
  chunks: Chunk[],
  embedding: EmbeddingConfig,
  full: boolean,
): Promise<boolean> {
  const [kind, endpoint, model] = normalizedBackend(embedding);
  const embedder = makeEmbedder(kind, endpoint, model, process.env["OPENAI_API_KEY"] ?? "");
  const store = new VectorStore(join(bp, "t2", "lancedb"));

  const docPrefix = embedding.document_prefix;
  const queryPrefix = embedding.query_prefix;
  const old = readJson(join(bp, "t2", "embedding.json"));
  let sameBackend =
    !full &&
    old !== null &&
    old["kind"] === kind &&
    old["endpoint"] === endpoint &&
    old["model"] === model &&
    (old["document_prefix"] ?? "") === docPrefix &&
    (old["query_prefix"] ?? "") === queryPrefix;

  let toEmbed: Chunk[];
  let deleteIds: Set<string>;
  if (sameBackend) {
    const embedded = await store.existingShas();
    const newIds = new Set(chunks.map((c) => c.id));
    toEmbed = chunks.filter((c) => embedded.get(c.id) !== c.sha256);
    deleteIds = new Set<string>();
    for (const id of embedded.keys()) if (!newIds.has(id)) deleteIds.add(id);
    for (const c of toEmbed) if (embedded.has(c.id)) deleteIds.add(c.id);
  } else {
    toEmbed = chunks;
    deleteIds = new Set();
  }

  let vectors = toEmbed.length > 0 ? await embedder.embed(toEmbed.map((c) => docPrefix + c.text)) : [];
  let dim: number;
  if (toEmbed.length > 0) dim = vectors[0]!.length;
  else if (sameBackend) dim = Math.trunc(Number(old!["dim"]));
  else if (embedding.dim) dim = Math.trunc(embedding.dim);
  else dim = (await embedder.embed(["brainpick"]))[0]!.length; // discover once, even with no chunks

  if (sameBackend && dim !== Math.trunc(Number(old!["dim"]))) {
    // the backend answers with a new dimensionality — every old vector is invalid
    sameBackend = false;
    toEmbed = chunks;
    vectors = await embedder.embed(chunks.map((c) => docPrefix + c.text));
  }

  const rows: ChunkRow[] = toEmbed.map((c, i) => ({
    id: c.id,
    doc: c.doc,
    ord: c.ord,
    text: c.text,
    vector: vectors[i]!,
  }));
  if (sameBackend) {
    if (rows.length > 0 || deleteIds.size > 0) await store.upsert(rows, deleteIds, dim);
  } else {
    await store.replaceAll(rows, dim);
  }

  const record: Record<string, string | number> = {
    dim,
    endpoint,
    fingerprint: fingerprint(kind, endpoint, model, dim, docPrefix, queryPrefix),
    kind,
    model,
  };
  if (docPrefix || queryPrefix) {
    // spec/30: keys appear only when a prefix is set
    record["document_prefix"] = docPrefix;
    record["query_prefix"] = queryPrefix;
  }
  return writeIfChanged(join(bp, "t2", "embedding.json"), canonicalJson(record));
}
